Engine/Module/ModuleSpacePartitioning.cpp

```diff
@@ -0,0 +1,267 @@
+#include "ModuleSpacePartitioning.h"
+
+#include "Main/Application.h"
+#include "Module/ModuleDebug.h"
+#include "Module/ModuleScene.h"
+#include "Module/ModuleRender.h"
+
+#include "Component/Component.h"
+#include "Component/ComponentMeshRenderer.h"
+
+#include <algorithm>
+#include <assimp/scene.h>
+#include <MathGeoLib.h>
+#include <Brofiler/Brofiler.h>
+
+bool ModuleSpacePartitioning::Init()
+{
+	ol_abbtree = std::make_unique<OLAABBTree>(INITIAL_SIZE_AABBTREE);
+	ol_quadtree = std::make_unique<OLQuadTree>();
+	ol_octtree = std::make_unique<OLOctTree>();
+
+	return true;
+}
+
+bool ModuleSpacePartitioning::CleanUp()
+{
+	return true;
+}
+
+void ModuleSpacePartitioning::GenerateQuadTree()
+{
+
+	BROFILER_CATEGORY("Generate QuadTree", Profiler::Color::Lavender);
+	AABB2D global_AABB;
+	global_AABB.SetNegativeInfinity();
+
+	for (const auto& mesh : App->renderer->meshes)
+	{
+		float minX = std::fmin(mesh->owner->aabb.bounding_box2D.minPoint.x, global_AABB.minPoint.x);
+		float minY = std::fmin(mesh->owner->aabb.bounding_box2D.minPoint.y, global_AABB.minPoint.y);
+
+		float maxX = std::fmax(mesh->owner->aabb.bounding_box2D.maxPoint.x, global_AABB.maxPoint.x);
+		float maxY = std::fmax(mesh->owner->aabb.bounding_box2D.maxPoint.y, global_AABB.maxPoint.y);
+		global_AABB.maxPoint = float2(maxX, maxY);
+		global_AABB.minPoint = float2(minX, minY);
+	}
+
+	ol_quadtree->Create(global_AABB);
+	for (const auto& mesh : App->renderer->meshes)
+	{
+		ol_quadtree->Insert(*mesh->owner);
+	}
+}
+
+void ModuleSpacePartitioning::GenerateOctTree()
+{
+
+	BROFILER_CATEGORY("Generate OctTree", Profiler::Color::Lavender);
+	AABB global_AABB;
+	global_AABB.SetNegativeInfinity();
+
+	for (const auto& mesh : App->renderer->meshes)
+	{
+		float minX = std::fmin(mesh->owner->aabb.bounding_box.minPoint.x, global_AABB.minPoint.x);
+		float minY = std::fmin(mesh->owner->aabb.bounding_box.minPoint.y, global_AABB.minPoint.y);
+		float minZ = std::fmin(mesh->owner->aabb.bounding_box.minPoint.z, global_AABB.minPoint.z);
+
+		float maxX = std::fmax(mesh->owner->aabb.bounding_box.maxPoint.x, global_AABB.maxPoint.x);
+		float maxY = std::fmax(mesh->owner->aabb.bounding_box.maxPoint.y, global_AABB.maxPoint.y);
+		float maxZ = std::fmax(mesh->owner->aabb.bounding_box.maxPoint.z, global_AABB.maxPoint.z);
+
+		global_AABB.maxPoint = float3(maxX, maxY, maxZ);
+		global_AABB.minPoint = float3(minX, minY, minZ);
+	}
+
+	ol_octtree->Create(global_AABB);
+	for (const auto&  mesh : App->renderer->meshes)
+	{
+		ol_octtree->Insert(*mesh->owner);
+	}
+}
+
+void ModuleSpacePartitioning::InsertAABBTree(GameObject* game_object)
```

Comment:
2 silly questions:
Why there is a Generate Octree, Generate Quadtree, but not AABB and the AABB have the functions here?

Also, the Octree and Quadtree are generated every frame?  If I insert a new GO, is all the octree generated from scratch?

Reply:
Hey, aabb tree is reset when changing scene but otherwise only is partially recomputed when moving/inserting/deleting elements.  But quadtree and octree are generated from scratch again every time. As are thought to be used for static objects and hence won't have changes in realtime i don't see it that bad. But maybe it is true that could be done in other ways.